sprint 14

kgontsov opened the pull request.

kgontsov committed:
final sprt 14

Nikitaaeee commented on Tracker/Controllers/CreateTrackerViewController.swift:
```diff
@@ -0,0 +1,204 @@
+        self.trackerOptions = options
+    }
+        
+    private func setupViews() {
+        [pageTitle, trackerNameTextField, trackerOptionsTableView, buttonStackView].forEach { view.falseAutoresizing($0) }
+        
+        NSLayoutConstraint.activate([
+            pageTitle.centerXAnchor.constraint(equalTo: view.centerXAnchor),
+            pageTitle.topAnchor.constraint(equalTo: view.topAnchor, constant: 27),
+            
+            trackerNameTextField.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 16),
+            trackerNameTextField.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -16),
+            trackerNameTextField.heightAnchor.constraint(equalToConstant: 75),
+            trackerNameTextField.topAnchor.constraint(equalTo: pageTitle.bottomAnchor, constant: 38),
+            
+            trackerOptionsTableView.topAnchor.constraint(equalTo: trackerNameTextField.bottomAnchor, constant: 24),
+            trackerOptionsTableView.leadingAnchor.constraint(equalTo: trackerNameTextField.leadingAnchor),
+            trackerOptionsTableView.trailingAnchor.constraint(equalTo: trackerNameTextField.trailingAnchor),
+            trackerOptionsTableView.heightAnchor.constraint(equalToConstant: CGFloat(trackerOptions.count * 75)),
+            
+            buttonStackView.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 16),
+            buttonStackView.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -16),
+            buttonStackView.bottomAnchor.constraint(equalTo: view.bottomAnchor, constant: -50),
+            buttonStackView.heightAnchor.constraint(equalToConstant: 60)
+        ])
+    }
```
- Было бы здорово делить настройку UI на два разных метода (например, _setupViews_ и _setupConstraints_).
В одном добавлять вью на супервью, а в другом настраивать констрейнты. 
Всегда лучше иметь несколько более мелких методов, чем один большой, выполняющий разные задачи.

- Название метода _falseAutoresizing_ в расширении UIView может быть немного запутанным, так как оно не явно указывает на то, что делает метод. 
Хорошее название метода должно быть интуитивно понятным и отражать его функциональность. 
Так как метод добавляет View на супервью и отключает autoresizingMasksIntoConstraints, можно назвать его, например, `addSubviewWithAutoLayout(_ view: UIView)`, или  `addSubviewAndDisableAutoresizing(_ view: UIView)`
Эти названия уже лучше указывают на то, что делает метод.

Nikitaaeee commented on Tracker/Controllers/CreateTrackerViewController.swift:
```diff
@@ -0,0 +1,204 @@
+        self.presentingViewController?.presentingViewController?.dismiss(animated: true)
+    }
+}
+
+// MARK: - UITableViewDataSource
+
+extension CreateTrackerViewController: UITableViewDataSource {
```
Круто, что ты используешь комментарии в стиле `// MARK: - UITableViewDataSource` ! 🔥

Было бы здорово добавлять их и к другим участкам кода.
Пример использования этих комментариев для организации кода можно [посмотреть тут](https://habr.com/ru/articles/744470/).

Nikitaaeee commented on Tracker/Controllers/CreateTrackerViewController.swift:
```diff
@@ -0,0 +1,204 @@
+        guard let weekSchedule = weekSchedule else { return }
+        self.weekSchedule = weekSchedule
+    }
+}
```
Тут, например, можно еще добавить `// MARK: - UITextFieldDelegate` и `// MARK: - ScheduleProtocolDelegate `

Nikitaaeee commented on Tracker/Controllers/TrackersViewController.swift:
```diff
@@ -0,0 +1,305 @@
+    
+    private let placeholderView = EmptyView(
+        title: "Что будем отслеживать?"
+    )
+    
+    
+    init() {
```
Можно удалить одну лишнюю пустую строку, одной достаточно:)

Nikitaaeee commented on Tracker/Structs/TrackerModels.swift:
```diff
@@ -0,0 +1,83 @@
+
+import UIKit
+
+enum WeekDay: CaseIterable {
+    case sunday
+    case monday
+    case tuesday
+    case wednesday
+    case thursday
+    case friday
+    case saturday
+    
+    var numberValue: Int {
+        switch self {
+        case .sunday:
+            return 1
+        case .monday:
+            return 2
+        case .tuesday:
+            return 3
+        case .wednesday:
+            return 4
+        case .thursday:
+            return 5
+        case .friday:
+            return 6
+        case .saturday:
+            return 7
+        }
+    }
```
Можно немного уменьшить количество кода, воспользовавшись удобной штукой, которую нам предоставляют enum'ы в  **Swift**:
- Подписать перечисление на **Int**
- Задать `rawValue` для каждого из кейсов
При этом не обязательно писать 
 case sunday = 1
 case monday = 2
 case tuesday = 3
Достаточно указать начальное число, а все остальные кейсы возьмут себе следующие числа. ⚡️

Вот так это будет выглядеть в коде:
```
enum WeekDay: Int, CaseIterable {
    case sunday = 1
    case monday
    case tuesday
    case wednesday
    case thursday
    case friday
    case saturday

    var numberValue: Int {
        return self.rawValue
    }
    ... остальной код ...
}
```

kgontsov committed:
Исправление замечаний ревью.
ps: локально у меня файл logo в ассетах с маленькой буквы, но не первый раз такой баг при заливке на гит. На ранних спринтах бывало такое с кастомными цветами, после пуша они оказывались с большой буквы ¯\_(ツ)_/¯
